test-app/tests/integration/components/headless-form-test.gts

```diff
@@ -494,6 +514,837 @@ module('Integration Component headless-form', function (hooks) {
+                @validate={{formValidateCallback}}
+                as |form|
+              >
+                <form.field
+                  @name="firstName"
+                  @validate={{fieldValidateCallback}}
+                  as |field|
+                >
+                  <field.label>First Name</field.label>
+                  <field.input data-test-first-name />
+                  <field.errors data-test-first-name-errors as |errors|>
+                    {{#each errors as |e index|}}
+                      <div data-test-error={{index}}>
+                        <div data-test-error-type>
+                          {{e.type}}
+                        </div>
+                        <div data-test-error-value>
+                          {{e.value}}
+                        </div>
+                        <div data-test-error-message>
+                          {{e.message}}
+                        </div>
+                      </div>
+                    {{/each}}
+                  </field.errors>
+                </form.field>
+                <form.field @name="lastName" as |field|>
+                  <field.label>Last Name</field.label>
+                  <field.input data-test-last-name />
+                  <field.errors data-test-last-name-errors />
+                </form.field>
+                <button type="submit" data-test-submit>Submit</button>
+              </HeadlessForm>
+            </template>);
+
+            await click('[data-test-submit]');
+
+            assert
+              .dom('[data-test-first-name-errors] [data-test-error]')
+              .exists({ count: 2 });
+
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-type]'
+              )
+              .hasText('uppercase');
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-value]'
+              )
+              .hasText('foo');
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="0"] [data-test-error-message]'
+              )
+              .hasText('First name must be upper case!');
+
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-type]'
+              )
+              .hasText('notFoo');
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-value]'
+              )
+              .hasText('foo');
+            assert
+              .dom(
+                '[data-test-first-name-errors] [data-test-error="1"] [data-test-error-message]'
+              )
+              .hasText('Foo is an invalid first name!');
+
+            assert.dom('[data-test-last-name-errors]').doesNotExist();
+          });
+        })
+      );
+    });
+  });
+
   module('Glint', function () {
```

Comment:
might be worth it to extract this to a separate file -- these files are getting quite long, and when it comes time to do any sort of automated migration on these files, having them be a smidge smaller allows for better incremental migration.

Reply:
Yeah, I wondered about this too, and had no clear idea how to structure this best.

So, when following the existing Ember convention of having `component/foo-test.js` for component `foo`, then this is what you end up, as `<HeadlessForm>` is the only public component the addon exposes. In other projects (e.g. ember-bootstrap) I had tests for the yielded components in separate files, like here it could be `components/headless-form/field.gts`, but neither directly accessing the component as `<HeadlessForm::Field>` (which wouldn't work currently, and be pretty terrible actually as that is not testing the real API) no putting tests that use `<HeadlessForm as |form|>` into a test file that is specific for the sub-components felt right.

But yeah, I think we don't have to follow the convention strictly here, and splitting this up into separate features of the same component as you suggested (like basics, validation, glint) would make sense. Will do that!